### PR TITLE
fix(HierarchyList): scroll only if needed and support secondary nodes

### PR DESCRIPTION
**Pull request by sstone2423**
**Summary**

- Scrolling was always happening on item selection regardless of whether the item was already in view or not causing some weird jumping
- `secondaryValue` was not accepting nodes in the list

**Change List (commits, features, bugs, etc)**

- Added the `scroll-into-view-if-needed` package to check if the node is visible or not, before attempting to scroll to it
- Allowed `secondaryValue` to accept nodes
- Added string type check for searching secondary values
- Updated 1 HierarchyList story to use a custom component as `secondaryValue`

**Acceptance Test (how to verify the PR)**

- Go to /?path=/story/watson-iot-hierarchylist--with-default-selected-id in chrome, firefox, and safari and ensure the item is scrolled to on initial render, then continue to select other items and ensure no scrolling occurs
- Go to /?path=/story/watson-iot-hierarchylist--with-mixed-hierarchies, open the 'Tasks' category, and see that the InlineLoading component is successfully rendering, then use the search bar and ensure searching still works (this ensures I didn't break the search by adding the node)

### CODE REVIEW

File: packages/react/src/components/List/HierarchyList/HierarchyList.jsx [hunk 1]
```diff
@@ -206,7 +208,7 @@ const HierarchyList = ({
 }) => {
   const [expandedIds, setExpandedIds] = useState(defaultExpandedIds);
   const [searchValue, setSearchValue] = useState('');
-  const [filteredItems, setFilteredItems] = useState(cloneDeep(items));
+  const [filteredItems, setFilteredItems] = useState(() => cloneDeep(items));
```
Comment (scottdickerson) [inline packages/react/src/components/List/HierarchyList/HierarchyList.jsx:211]: This is very smart!

**Review by scottdickerson — APPROVED** [review]
LGTM

File: packages/react/src/components/List/HierarchyList/HierarchyList.jsx (hunk 1 above)
Comment (davidicus) [reply packages/react/src/components/List/HierarchyList/HierarchyList.jsx:211]: I definitely think it's good to instantiate the state this way but what is the reason for the cloneDeep? Especially now.

File: packages/react/src/components/List/HierarchyList/HierarchyList.jsx [hunk 2]
```diff
@@ -40,7 +41,7 @@ const propTypes = {
   /** Buttons to be presented in List header */
   buttons: PropTypes.arrayOf(PropTypes.node),
   /** ListItems to be displayed */
-  items: PropTypes.arrayOf(PropTypes.any).isRequired,
+  items: PropTypes.arrayOf(PropTypes.shape(ListItemPropTypes)),
```
Comment (davidicus) [inline packages/react/src/components/List/HierarchyList/HierarchyList.jsx:44]: This is a breaking change of an item that we just took out of experimental. We should make this an enum with the deprecation warning for the old way included. I see it's no longer required so what behavior is changed by this new shape?

**Review by davidicus — CHANGES_REQUESTED** [review]
We need continue to support the old prop type since this will be in a minor release

File: packages/react/src/components/List/HierarchyList/HierarchyList.jsx (hunk 1 above)
Comment (sstone2423) [reply packages/react/src/components/List/HierarchyList/HierarchyList.jsx:211]: This is to keep the prop items intact as they'll get mutated a bunch inside of the component

File: packages/react/src/components/List/HierarchyList/HierarchyList.jsx (hunk 2 above)
Comment (sstone2423) [reply packages/react/src/components/List/HierarchyList/HierarchyList.jsx:44]: This enforces the same prop types as the List below expects so the `any` was always wrong and would throw prop types warnings regardless

Comment (davidicus) [reply packages/react/src/components/List/HierarchyList/HierarchyList.jsx:44]: Yeah I see down in the list now. Thanks.

File: packages/react/src/components/List/HierarchyList/HierarchyList.jsx (hunk 1 above)
Comment (sstone2423) [reply packages/react/src/components/List/HierarchyList/HierarchyList.jsx]: Removed as its actually not needed

**Review by davidicus — APPROVED** [review]

**Pull request merged**